add .pre-commit-config.yaml

sfc-gh-tszerszen opened the pull request:
<!--
Before contributing (PLEASE READ!)

⚠️ If your contribution is more than a few lines of code, then prior to starting to code on it please post in the issue saying you want to volunteer, then wait for a positive response. And if there is no issue for it yet, create it first.

This helps make sure:

  1. Two people aren't working on the same thing
  2. This is something Streamlit's maintainers believe should be implemented/fixed
  3. Any API, UI, or deeper architectural changes that need to be implemented have been fully thought through by Streamlit's maintainers
  4. Your time is well spent!

More information in our wiki: https://github.com/streamlit/streamlit/wiki/Contributing
-->

## 📚 Context

_Please describe the project or issue background here_

- What kind of change does this PR introduce?

  - [ ] Bugfix
  - [x] Feature
  - [ ] Refactoring
  - [ ] Other, please describe:

## 🧠 Description of Changes

The goal of this PR is to get rid of `husky` and use `pre-commit` as our tool for git hooks.

This PR removes `husky` dependency from our `packages.json` and updates both lock files.
I've updated [Contributing page](https://github.com/streamlit/streamlit/wiki/Contributing) to install `pre-commit` during setup of the project.

On commands `make all-devel / make mini-devel / make init` command `pre-commit install` is run.

This PR creates `.pre-commit-config.yaml` configuration file, which can be used by `pre-commit`.
The `.pre-commit-config.yaml` is configured to run automatically `black` and `prettier` on all commited files before commiting the change.

### Pre-commit usage
```bash
# make sure you have pre-commit installed 
# as described on Contributing page 
# https://github.com/streamlit/streamlit/wiki/Contributing

# run one of below commands
$ make all-devel / make mini-devel / make init / make pre-commit-install

# make some changes inside our Python or Javascript code
$ git commit -am "your commit message"

# you should see black/prettier being run
```

- _Add bullet points summarizing your changes here_

  - [ ] This is a breaking API change
  - [ ] This is a visible (user-facing) change

**Revised:**

_Insert screenshot of your updated UI/code here_

**Current:**

_Insert screenshot of existing UI/code here_

## 🧪 Testing Done

- [ ] Screenshots included
- [ ] Added/Updated unit tests
- [ ] Added/Updated e2e tests

## 🌐 References

_Does this depend on other work, documents, or tickets?_

- **Issue**: Closes #XXXX

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-tszerszen committed:
add .pre-commit-config.yaml

sfc-gh-kbregula commented:
Should we add docs about it to contributor guide? 
Here is an example of docs from Apache Airflow project: 
https://github.com/apache/airflow/blob/main/STATIC_CODE_CHECKS.rst#pre-commit-hooks

sfc-gh-kbregula commented:
Should we also run pre-commit on CI?

sfc-gh-tszerszen commented:
@sfc-gh-kbregula I strongly agree on adding more steps to CI 👍 

When it comes to Contribution guideline, I think we need Single source of truth about pre-commits. If we use `husky` and if it works, then we should stick to one tool to not introduce mess and miscommunication. That's why I'm waiting for more feedback especially how to setup `husky` and if it meets the needs we commonly agree on.

Once we agree, as a team, how we will handle git hooks I will update contribution guideline accordingly

sfc-gh-tszerszen committed:
remove husky

sfc-gh-tszerszen committed:
add prettier and make pre-commit-install

sfc-gh-kbregula commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,10 @@
+    hooks:
+      - id: black
+        language_version: python3.8
+  - repo: https://github.com/pre-commit/mirrors-prettier
```
Should we also add eslint?
https://github.com/streamlit/streamlit/blob/fbcd897ba5d68a0016a2d087e75c5209f85dc765/Makefile#L270-L294
and typecheck?
https://github.com/streamlit/streamlit/blob/fbcd897ba5d68a0016a2d087e75c5209f85dc765/Makefile#L296-L299

sfc-gh-tszerszen replied:
@sfc-gh-kbregula thank you for pointing this out 👍 I think we should add both `typecheck` and `eslint`, what's more we should think about adding `mypy` script. I've just tried to do it, and:

- didn't found ready pre-commit fork of `typecheck` we must do it with custom command
- there is eslint precommit fork, however it didn't discover configuration file `.eslintrc` which is not in the root directory of the project, so I must spend more time the ways how to fix it
- mypy requires using our pipenv shell and our custom mypy scripts

My proposition is to merge this PR, and step by step add more to our pre-commit script, is that ok for you?

sfc-gh-kbregula replied:
I'm fine with this. This is still a step forward.

sfc-gh-tszerszen replied:
@sfc-gh-kbregula thank you for the help with pre-commit.yaml I've updated this PR with 
* ~isort~ isort will modify a lot of files, for that reason I've removed it from this PR
* typecheck
* eslint
Can you double check if these changes are ok? I've added `files` selector to every hook

sfc-gh-tszerszen committed:
do not install pre-commit on circleci

sfc-gh-tszerszen committed:
add pre-commit autoupdate to makefile

sfc-gh-tszerszen committed:
add typecheck

sfc-gh-tszerszen committed:
add eslint

sfc-gh-tszerszen committed:
add eslint

sfc-gh-tszerszen committed:
add isort

sfc-gh-tszerszen committed:
add files checks

sfc-gh-tszerszen committed:
remove isort because of lots of file to change

lukasmasuch commented on Makefile:
```diff
@@ -21,7 +21,7 @@ all: init frontend install
 .PHONY: all-devel
 # Get dependencies and install Streamlit into Python environment -- but do not build the frontend.
-all-devel: init develop
+all-devel: init develop pre-commit-install
```
I don't think we need to call `pre-commit-install` since it is already part of `init`, or? On the other hand, maybe we even remove it from `init` and only have it in our `devel` (`all-devel` and `mini-devel`) calls since this is clearly a development targeted dependency.

vdonato replied:
+1 to calling `pre-commit-install` in the two devel targets and not `init`

sfc-gh-tszerszen replied:
@LukasMasuch @vdonato good point 👍 I've just changed it 👍

lukasmasuch commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,26 @@
+        pass_filenames: false
+      - id: run-eslint
+        name: Run run eslint
+        entry: cd frontend && ./frontend/node_modules/.bin/eslint --ext .js --ext .jsx --ext .ts --ext .tsx --ignore-pattern './frontend/src/autogen/*' --max-warnings 0 ./frontend/src
```
Would it be possible here if we just call `make jslint`, otherwise we are kind of duplicating this logic?

sfc-gh-kbregula replied:
pre-commit only runs on files that are changed i.e. when entry equals `black` and the files changed are `a.py` and `b.py`, the pre-commit executes the command: `black a.py b.py`.

lukasmasuch replied:
ahh I see 👍  So the difference is that the `make jslint` runs on all files and the eslint command with pre-commit only runs on files that have been changed in the commit?

sfc-gh-kbregula replied:
Yes. Pre-commit is a little smareter and quicker.

lukasmasuch commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,26 @@
+        files: \.js$|\.jsx$|\.ts$|\.tsx$
+        language: node
+        pass_filenames: false
+      - id: run-eslint
```
What happens if the linter detects some problems? Is the commit aborted? Is there an error message that is shown to the user even when git commit is not triggered from the terminal?

sfc-gh-tszerszen replied:
@LukasMasuch thanks for your question, it reminded me that I should add some explanation both in the code and [Contributing page](https://github.com/streamlit/streamlit/wiki/Contributing). I've added **Pre-commit hooks** section into [Contributing page](https://github.com/streamlit/streamlit/wiki/Contributing) and some comments into `.pre-commit-config.yaml`.

Here is my comment:
```
# Pre-commit configuration file,
# when Streamlit's pre-commit detects that one of the linters has failed,
# it automatically lints the files and does not allow the commit to pass. 
# Please review the changes after lint has failed and commit them again,
# the second commit should pass,
# because the files were linted after trying to do the first commit.
```

When you use IDE for example: VSCode it will display a message with git logs and info that commit has failed 👍

lukasmasuch reviewed (commented):
Overall looks good 👍  I added some initial comments and questions

vdonato reviewed (commented):
LGTM, although I'll leave it to a person with dependency-approval powers to +1 this since my review won't unblock merge anyway 🙂

sfc-gh-tszerszen committed:
update prettier and work on eslint

sfc-gh-kbregula committed:
Add workaround for hooks in subdirectory

sfc-gh-kbregula committed:
fixup! Add workaround for hooks in subdirectory

sfc-gh-kbregula commented on Makefile:
```diff
@@ -253,43 +249,19 @@ react-build:
-			--ignore-pattern 'src/autogen/*' \
-			--max-warnings 0 \
-			--format junit \
-			--output-file test-reports/eslint/eslint.xml \
```
Is this file in use anywhere?

sfc-gh-kbregula committed:
fixup! fixup! Add workaround for hooks in subdirectory

sfc-gh-kbregula committed:
fixup! fixup! fixup! Add workaround for hooks in subdirectory

sfc-gh-kbregula committed:
fixup! fixup! fixup! fixup! Add workaround for hooks in subdirectory

sfc-gh-kbregula committed:
fixup! fixup! fixup! fixup! fixup! Add workaround for hooks in subdirectory

sfc-gh-kbregula commented:
I did some work on it and made some improvements:
- I've added the `script/run_in_subdirectory.py` file to work around the pre-commit problem with running hooks in subdirectories. See: https://github.com/pre-commit/pre-commit/issues/1417
- pre-commit is run on the CI to make sure its configuration is error free and to avoid regression.

sfc-gh-kbregula committed:
fixup! fixup! fixup! fixup! fixup! fixup! Add workaround for hooks in subdirectory

sfc-gh-kbregula committed:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! Add workaround for hooks in subdirectory

sfc-gh-kbregula committed:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! Add workaround for hooks in subdirectory

sfc-gh-kbregula committed:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! Add workaround for hooks in subdirectory

sfc-gh-kbregula committed:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! Add workaround for hooks in subdirectory

sfc-gh-tszerszen commented:
@sfc-gh-kbregula thank you because I got stuck with this PR 👍 You explained me a lot about pre-commit and have done awesome job here ❤️ 

I'm worried we've added pre-commit to our CI though, I think this approach is better, since it makes sure we're good with the introduced changes, but we need to reconfigure our CI. @LukasMasuch what do you think about current state of the PR and ~can we reconfigure the circle CI to have pre-commit installed? (who should do this? and if you could help how? I've never configured circle CI myself, but I'm happy to learn)~
**CircleCI works so this point does not matter**

@LukasMasuch @vdonato thank you for your suggestions, I've added them to this PR 👍

sfc-gh-tszerszen committed:
remove pre-commit-install from init makefile

sfc-gh-tszerszen committed:
add instructions about how pre-commit works

vdonato commented on .circleci/config.yml:
```diff
@@ -496,7 +497,6 @@ jobs:
       - run:
           name: Run type checkers
           command: |
-            make tstypecheck
             scripts/mypy --report
```
nit: Would be useful to have a comment here noting that `tsc` is run by `pre-commit`

sfc-gh-tszerszen replied:
Added a comment 👍

vdonato commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,40 @@
+      - id: black
+        args:
+          # Configure Black to support only syntax supported by the minimum supported Python version in setup.py.
+          - --target-version=py36
```
nit: I think our min python version now is 3.7 (the version in the Makefile must have been set when it was min 3.6)

sfc-gh-tszerszen replied:
Thank you, it's changed 👍

vdonato reviewed (approved):
Found two small things, but otherwise LGTM!

This will still need approval from someone with dependency-approval powers as it made some changes to our `package.json`. The approval should just be a formality, though, as we only removed packages.

sfc-gh-tszerszen committed:
add comment about tsc being run by pre-commit and change black target version to py37

lukasmasuch commented on .circleci/config.yml:
```diff
@@ -486,17 +486,17 @@ jobs:
+          name: Run type checkers # tsc is run by pre-commit, thus in this step we only run mypy
           command: |
-            make tstypecheck
             scripts/mypy --report
```
nit: Maybe we also just move this to the `Run linters` step and remove `Run type checkers` step since we anyways already do the typescript type checking in that step.

lukasmasuch commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,40 @@
+        pass_filenames: true
+      - id: typecheck
+        name: Typecheck
+        entry: yarn --cwd frontend tsc --noEmit
```
Is `./scripts/run_in_subdirectory.py frontend/ ` not needed for the typecheck?

lukasmasuch replied:
btw. there is also a `typecheck` command defined in the `package.json`

sfc-gh-kbregula replied:
Not needed because the `pass_filenames` flag is set to `False`, I have had trouble configuring tsc execution to run on individual files only, so I set `pass_filenames` to `False`. This is minimally slower as all files are checked, but at least it works without errors. The current behavior is identical to the behavior from main branch, so we have no performance regression.

lukasmasuch reviewed (approved):
LGTM 👍  It would be good if you could also summarize (and share in Slack) what people need to do with their existing local setup once this is merged into `develop`. I assume this includes installation of `pre-commit` and maybe running `all-devel` again? 

Also, @willhuang1997 and @maryharges are working on a migration from CircleCI to Github Action (see this [PR](https://github.com/streamlit/streamlit/pull/5079)). It would be great if you could reach out to them after merging to get pre-commit into the Github Action setup as well.

jroes reviewed (approved):
Not seeing any new dependencies here, approved from a dependency / licensing POV

sfc-gh-kbregula commented:
🎉
